courses/static_analysis_via_compiler/010_static_analysis_via_compiler.rst

```diff
@@ -0,0 +1,1479 @@
+  + High integrity restrictions (see H.4)
+
++ GNAT defines additional restriction identifiers
++ All restrictions, both language-defined and GNAT-defined, are listed and described in the *GNAT Reference Manual*
+
+------------------------
+Restriction Categories
+------------------------
+
++ Portability
++ Allocation
++ Access Types & Values
++ Exceptions
++ OOP
++ Tasking
++ Real-Time Programming
++ Code Generation
++ Miscellaneous
++ GNAT defines additional restrictions in all these categories
+
+   + We examine some of them here...
+
+----------------------------------
+Applying Restriction Identifiers
+----------------------------------
+
++ In source or in configuration file
+
+  + Configuration file name can be default or given by gpr file
+
+     + Default config file name is :filename:`gnat.adc`
+
+  .. code:: Ada
+
+     pragma Restrictions (No_Implicit_Heap_Allocations);
+     pragma Restrictions (No_Implicit_Conditionals);
+     pragma Restrictions (No_Entry_Calls_In_Elaboration_Code);
+
++ :toolname:`GNATbind` can list all restrictions that could be applied to the code corresponding to a given ALI file
+
+  + Via :command:`-r` switch
+  + Useful for code audit, and code generation control
+
+------------------
+OOP Restrictions
+------------------
+
++ :ada:`No_Dispatch` (RM H.4)
+
+  + Ensures no occurrences of :ada:`T'Class` for any tagged type :ada:`T`
+  + Prevents dynamic dispatching (but also other usage)
+
++ :ada:`No_Dispatching_Calls` (GNAT)
+
+  + Ensures generated code involves no dispatching calls
+  + Allows
+
+    + Record extensions
+    + Classwide membership tests
+    + Other classwide features
+
+  + Does not allow involving implicit dispatching
+
+  + Comparable to :ada:`No_Dispatch`
+
+    + Except allows all classwide constructs that do not imply dispatching
+
+------
+Quiz
+------
+
+.. container:: columns
+
+  .. container:: column
+
+    .. container:: latex_environment tiny
+
+      .. code:: Ada
+
+        package Definition is
+           type T is tagged record
+              Data : Natural;
+           end record;
+           procedure P (X : T);
+           type Dt is new T with record
+              More_Data : Natural;
+           end record;
+           not overriding procedure Q (X : Dt);
+        end Definition;
+
+      .. code:: Ada
+        :number-lines: 1
+
+        pragma Restrictions (No_Dispatching_Calls);
+
+        with Definition; use Definition;
+        procedure Demo (O : T'class) is
+           N : Natural := O'size;
+           C : T'class := O;
+        begin
+           if O in Dt'class then
+              Q (Dt (O));
+           else
+              P (O);
+           end if;
+        end Demo;
+
+  .. container:: column
+
+    Which line number(s) violate the restriction?
+
+    A. 5, 6, 8, 9, 11
+    B. 11
+    C. :answer:`5, 6, 11`
+    D. No violations
+
+    .. container:: animate
+
+       + Line 5 - Dispatch needed to determine size of O
+       + Line 6 - Just a copy
```

Comment:
I believe this is a bit weirdly phrased, getting the value from the concrete object would need solving T's concrete type, so somewhat "dispatching". The "just a copy" does not make it seem like there is dispatching involved

Reply:
Cleaned up the language